### PR TITLE
Query: Add optimization to conditional being equated to equality

### LINKED ISSUE #12148: Query: Property access on complex expression is not translated
Edit - updated reproduction without AutoMapper

```
Unhandled Exception: System.ArgumentNullException: Value cannot be null.
Parameter name: querySource
   at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.FindEntityType(IQuerySource querySource)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.NullCheckRemovalTestingVisitor.<AnalyzeTestExpression>b__7_0(IProperty p, IQuerySource qs)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.<>c__DisplayClass95_0.<BindMethodCallExpression>b__0(IProperty property, IQuerySource querySource)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.<>c__DisplayClass92_0`1.<BindMethodCallExpression>b__0(IReadOnlyList`1 ps, IQuerySource qs)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindPropertyExpressionCore[TResult](Expression propertyExpression, IQuerySource querySource, Func`3 propertyBinder)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindMethodCallExpression[TResult](MethodCallExpression methodCallExpression, IQuerySource querySource, Func`3 methodCallBinder)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindMethodCallExpression(MethodCallExpression methodCallExpression, Action`2 methodCallBinder)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.NullCheckRemovalTestingVisitor.AnalyzeTestExpression(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.NullCheckRemovalTestingVisitor.CanRemoveNullCheck(Expression testExpression, Expression resultExpression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.VisitConditional(ConditionalExpression expression)
   at System.Linq.Expressions.ConditionalExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.VisitMemberInit(MemberInitExpression memberInitExpression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalProjectionExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Remotion.Linq.Clauses.SelectClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateAsyncQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileAsyncQuery[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileAsyncQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass22_0`1.<CompileAsyncQuery>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddAsyncQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileAsyncQuery[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.System.Collections.Generic.IAsyncEnumerable<TResult>.GetEnumerator()
   at System.Linq.AsyncEnumerable.Aggregate_[TSource,TAccumulate,TResult](IAsyncEnumerable`1 source, TAccumulate seed, Func`3 accumulator, Func`2 resultSelector, CancellationToken cancellationToken)
```

```c#
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Console;

namespace EFCoreBugs
{
    class Program
    {
        static async Task Main(string[] args)
        {
            using (var db = new TestDb())
            {
                await db.Database.EnsureDeletedAsync();
                await db.Database.EnsureCreatedAsync();
            }
            using (var db = new TestDb())
            {
                await db.Blogs
                    .Select(dtoBlog => new BlogDto()
                    {
                        Post = (dtoBlog.Posts.Any() ? dtoBlog.Posts.FirstOrDefault() : null) == null ? null : new PostDto()
                    })
                    .ToListAsync();
            }

            Console.ReadLine();
        }
    }

    public class TestDb : DbContext
    {
        public DbSet<Blog> Blogs { get; set; }
        public DbSet<Post> Posts { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var log = new LoggerFactory(new[] { new ConsoleLoggerProvider((_, __) => true, true) });
            optionsBuilder.UseLoggerFactory(log);
            optionsBuilder.UseSqlServer("Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=master;Database=bugtest;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=True;ApplicationIntent=ReadWrite;MultiSubnetFailover=False");
        }
    }

    public class Blog
    {
        public Guid Id { get; set; }

        [InverseProperty(nameof(Post.Blog))]
        public virtual List<Post> Posts { get; } = new List<Post>();
    }
    public class Post
    {
        public Guid Id { get; set; }
        public Guid BlogId { get; set; }

        [ForeignKey(nameof(BlogId))]
        public virtual Blog Blog { get; private set; }
    }

    public class BlogDto
    {
        public PostDto Post { get; set; }
    }
    public class PostDto { }
}
```

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.0" />
  </ItemGroup>
</Project>
```

### Further technical details
EF Core version: 2.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system:  win10 x64
IDE: Visual Studio 2017 Preview 15.8.0 Preview 1.1

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #12148

### CODE REVIEW

File: src/EFCore/Query/Internal/NullCheckRemovingExpressionVisitor.cs [hunk 1]
```diff
@@ -18,6 +18,19 @@ public class NullCheckRemovingExpressionVisitor : ExpressionVisitor
...
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        protected override Expression VisitBinary(BinaryExpression binaryExpression)
+        {
+            var visitedExpression = base.VisitBinary(binaryExpression);
+
+            return TryOptimizeConditionalEquality(visitedExpression) ?? visitedExpression;
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/NullCheckRemovingExpressionVisitor.cs:31]: For now I added it here, we need to have 2 kinds of optimizations, before nav expansion and after nav expansion. I will revisit that in next release. (Possibly combining subquery member pushdown/nullremoval in one visitor)

File: src/EFCore/Query/Internal/NullCheckRemovingExpressionVisitor.cs [hunk 2]
```diff
@@ -60,6 +73,42 @@ protected override Expression VisitConditional(ConditionalExpression conditional
             return base.VisitConditional(conditionalExpression);
         }
 
+        private Expression TryOptimizeConditionalEquality(Expression expression)
+        {
+            // Simplify (a ? b : null) == null => a
+            // Simplify (a ? null : c) == null => !a
```
Comment (maumar) [inline src/EFCore/Query/Internal/NullCheckRemovingExpressionVisitor.cs:79]: what if b happens to be null as well?

Comment (maumar) [reply src/EFCore/Query/Internal/NullCheckRemovingExpressionVisitor.cs:79]: seems like the optimization should be `a && b != null` for the first and `!a || b == null` for the second

Comment (smitpatel) [reply src/EFCore/Query/Internal/NullCheckRemovingExpressionVisitor.cs:79]: // Simplify (a ? b : null) == null => !a || b == null
            // Simplify (a ? null : b) == null => a || b == null

truth table way

**Comment by smitpatel** [comment]
@Pilchie for RC1 approval.

**Comment by Pilchie** [comment]
Approved for RC1

**Pull request merged**